PFW-1334 Move E-motor on FSENSOR error retry

gudnimg opened the pull request:
JIRA ticket: https://dev.prusa3d.com/browse/PFW-1334

When a FSENSOR error happens, the first step for the MK3S should be to stop any loading by stopping the E-motor.

From this point, the buttons should determine what the MK3S does next.

gudnimg committed:
Partial fix for PFW-1334

When a FSENSOR error happens, the first step for the MK3S should be to stop any loading by stopping the E-motor.

From this point, the buttons should determine what the MK3S does next.

gudnimg commented:
~~The first commit fixes the scenario for `FSENSOR DIDNT TRIGGER`, because the Retry button correctly restarts the state machine by unloading the filament and repeating the load. Specifically this is for loading, not unloading.~~

EDIT: Correction, the above commit only stops the motor from moving endlessly after an error appears. Unloading is not triggered and will be added in a later commit.

I have yet to verify the other errors.

gudnimg commented:
* `FSENSOR TOO EARLY` is triggered during loading. If the FINDA is set to 1 during load as normal but the filament sensor is set 1 too quickly then this error is raised. A retry should unload to FINDA.

* `FSENSOR DIDNT GO OFF` is a bit more tricky. Because here we are not loading, but unloading. I am able to trigger this error if the retry for `FSENSOR TOO EARLY` fails. For whatever reason when retry is selected here, there is no U0 sent to the MMU like with `FSENSOR DIDNT TRIGGER`. 🤔

gudnimg committed:
PFW-1334 Move E-motor on FSENSOR error retry

gudnimg commented:
I will likely now need to distinguish between the normal call to `unload()` and a unload retry since they do not behave the same.

gudnimg committed:
Add MMU2_UNLOAD_TO_FINDA_FEED_RATE

gudnimg committed:
PFW-1334 Distinguish between a retry unload and U0 unload

U0 unload uses the ramming sequence, while the retry unload does not because the MMU starts unloading much sooner

gudnimg commented:
I'm marking this as ready for review, it should be safe to use now.

gudnimg committed:
Merge branch 'mk3-mmu' of https://github.com/DRracer/Prusa-Firmware into mk3-mmu

gudnimg commented:
Rebased branch onto mk3-mmu

3d-gussner commented on Firmware/mmu2.cpp:
```diff
@@ -765,6 +782,26 @@ void MMU2::OnMMUProgressMsg(ProgressCode pc){
+            st_synchronize();
+            unloadFilamentStarted = true;
+            // Unload slowly while MMU is initalising its axis
+            current_position[E_AXIS] -= 10.0f;
```
Why do we have hard coded values?

3d-gussner commented on Firmware/mmu2.cpp:
```diff
@@ -765,6 +782,26 @@ void MMU2::OnMMUProgressMsg(ProgressCode pc){
+            plan_buffer_line_curposXYZE(10.0f);
+            st_synchronize();
+            // Now do a fast unload in sync with the MMU
+            current_position[E_AXIS] -= 427.0f - 42.85f - 20.0f; // Roughly same distance as MMU plans
```
Same as above

3d-gussner commented on Firmware/mmu2.cpp:
```diff
@@ -777,6 +814,19 @@ void MMU2::OnMMUProgressMsg(ProgressCode pc){
+                
+                if (mmu2.FindaDetectsFilament() == 1)
+                { // Keep moving the E-motor until and error happens or FINDA untriggers
+                    current_position[E_AXIS] -= 6.0f;
```
Same as above

vintagepc commented:
I agree with @3d-gussner - if we must have magic numbers, they should at least be pulled to the top of the file with a static constexpr rather than scattered throughout the code. 

Of course, an improved solution with shared headers etc would be preferred.

gudnimg commented:
I will hopefully add a new commit later today with global constants and some documentation about them. While I was looking into documenting the constants I found that this is probably incorrect: `-= 427.0f - 42.85f - 20.0f` it should be `-= 427.0f + 42.85f + 20.0f`. I can verify this later today also.

This is basically this line from the MMU FW:  `mpu::pulley.PlanMove(-config::defaultBowdenLength - config::feedToFinda - config::filamentMinLoadedToMMU, config::pulleyUnloadFeedrate);`

gudnimg committed:
* Correct the unload length on the MK3S side
* Pull all numbers into global constants

gudnimg commented:
@3d-gussner @vintagepc I've pulled everything into constants. The variable names are a bit long I know 🙈 I tried to keep them descriptive and general.

The constant `-= 427.0f - 42.85f - 20.0f` is now corrected to `-= 427.0f + 42.85f + 20.0f` and it looks much better when executed. The MK3S is more in sync with the MMU. A new problem arose, `427.0f + 42.85f + 20.0f` is now considered too long of an extrusion so I've split this constant into two to keep the firmware happy.

gudnimg committed:
Add a missing break

I don't see any difference in behaviour but this should be more correct

gudnimg commented:
I've done some more testing and cannot see any issues with tool changes (during printing). I can also confirm the changes in the code are not blocking the MMU communication. So if the testers see the communication drop on this build, it must be something else unrelated to these changes.

The only issue left looking into is the reported layer shifts. I can confirm there aren't any gcodes being executing while loading or unloading (even during a retry). But one idea I have is we are skipping gcodes perhaps. I'm currently looking into how to log that to confirm it.
